src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionList.java

```diff
@@ -252,4 +229,53 @@ public void onFailure(Throwable t) {
         executor.shutdown();
         executor.awaitTermination(1, TimeUnit.MINUTES);
     }
+
+    /**
+     * Whenever a Pipeline resumes, resume all incomplete steps in its {@link FlowExecution}.
+     *
+     * Called by {@code WorkflowRun.onLoad}, so guaranteed to run if a Pipeline resumes regardless of its presence in
+     * {@link FlowExecutionList}.
+     */
+    @Extension
+    public static class ResumeStepExecutionListener extends FlowExecutionListener {
+        @Override
+        public void onResumed(@Nonnull FlowExecution e) {
+            Futures.addCallback(e.getCurrentExecutions(false), new FutureCallback<List<StepExecution>>() {
+                @Override
+                public void onSuccess(List<StepExecution> result) {
+                    if (e.isComplete()) {
+                        // WorkflowRun.onLoad will not fire onResumed if the serialized execution was already
+                        // complete when loaded, but right now (at least for workflow-cps), the execution resumes
+                        // asynchronously before WorkflowRun.onLoad completes, so it is possible that the execution
+                        // finishes before onResumed gets called.
+                        // That said, there is nothing to prevent the execution from completing right after we check
+                        // isComplete. If we want to fully prevent that, we would need to delay actual execution
+                        // resumption until WorkflowRun.onLoad completes or add some form of synchronization.
+                        return;
+                    }
+                    FlowExecutionList list = FlowExecutionList.get();
+                    FlowExecutionOwner owner = e.getOwner();
+                    if (!list.runningTasks.contains(owner)) {
+                        LOGGER.log(Level.WARNING, "Resuming {0}, which is missing from FlowExecutionList ({1}), so registering it now.",
+                                new Object[] {owner, list.runningTasks.getView()});
```

Comment:
Note that this is one of the unusual cases where the `Supplier` overload to `Logger.warning` would be inappropriate since the call to `getView` must happen now for the message to be meaningful.